### PR TITLE
Add async versions of `withSpan` API

### PR DESCRIPTION
**Pull request by slashmo**
Adds support for creating spans in an async context with automatic "baggage propagation" through task locals.

### CODE REVIEW

File: Sources/Tracing/Tracer.swift [hunk 1]
```diff
@@ -127,3 +127,65 @@ extension Tracer {
         }
     }
 }
+
+#if compiler(>=5.5) // we cannot write this on one line with `&&` because Swift 5.0 doesn't like it...
```
Comment (ktoso) [inline Sources/Tracing/Tracer.swift:131]: oh, lol okey :)

Comment (slashmo) [reply Sources/Tracing/Tracer.swift:131]: Copied directly from @weissi's NIO async/await support 😁

Comment (weissi) [reply Sources/Tracing/Tracer.swift:131]: do you support Swift 5.0 in this repo? If no, then you can remove this extra line, from Swift 5.1 on `#if compiler(>=5.5) && $AsyncAwait` works as expected. In Swift 5.0 that line will fail compilation and error (think it complains about something that has to be numeric or so).

**Review by ktoso — APPROVED** [review]
Fantastic, this'll be great 👍

File: Sources/Tracing/Tracer.swift (hunk 1 above)
Comment (ktoso) [reply Sources/Tracing/Tracer.swift:131]: For better or worse we currently do support 5.0, yes.

Though to be honest we'll drop it all and require async soon I guess...?
But for now yes, we do.

Comment (slashmo) [reply Sources/Tracing/Tracer.swift:131]: Yeah, I don't think we need to maintain backwards compatibility here.

Comment (weissi) [reply Sources/Tracing/Tracer.swift:131]: Given that this repo isn't 1.0.0 tagged, I'd recommend dropping all old Swift versions :P

Comment (ktoso) [reply Sources/Tracing/Tracer.swift:131]: Yeah, we'll do that.

The discussions last year during gsoc were on the dramatic side with people requesting 5.0... 😉 

Now as we want to require task-locals we'll rehash the lib and require async features anyway and drop the 5.0, not much to be discussed about it this year about it.

**Comment by slashmo** [comment]
Closed in favor of #52

**Pull request closed without merging**